File: apps/macos/LauncherApp/look-app/Views/Commands/CalcCommand.swift
import Foundation

enum CalcResult {
    case value(String)
    case error(String)
}

struct CalcCommand {
    static let maxMagnitude: Double = 1_000_000_000_000.0

    static func evaluate(_ expression: String) -> CalcResult {
        guard isReadyForEvaluation(expression) else {
            return .error("Invalid expression")
        }

        let normalized = decimalizeIntegerTokens(in: normalizeExpression(expression))

        do {
            var parser = Parser(input: normalized)
            let evaluated = try parser.parse()
            if abs(evaluated) > maxMagnitude {
                return .error("Error: result out of range (±1,000,000,000,000)")
            }
            return .value(formatFloat(evaluated))
        } catch ParserError.divisionByZero {
            return .error("Error: division by zero")
        } catch {
            return .error("Invalid expression")
        }
    }

    static func isReadyForEvaluation(_ expression: String) -> Bool {
        let trimmed = expression.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return false }

        var balance = 0
        for ch in trimmed {
            if ch == "(" { balance += 1 }
            if ch == ")" {
                balance -= 1
                if balance < 0 { return false }
            }
        }
        if balance != 0 { return false }

        if let last = trimmed.last, "+-*/^.(".contains(last) {
            return false
        }

        let allowedPattern = "^[0-9A-Za-z_+\\-*/%^!().:xXvV\\s]+$"
        guard let regex = try? NSRegularExpression(pattern: allowedPattern) else { return false }
        let full = NSRange(trimmed.startIndex..<trimmed.endIndex, in: trimmed)
        guard let match = regex.firstMatch(in: trimmed, range: full), match.range == full else {
            return false
        }
        return true
    }

    private static func formatFloat(_ value: Double) -> String {
        if value.isNaN || value.isInfinite { return "nan" }

        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.numberStyle = .decimal
        formatter.groupingSeparator = ","
        formatter.usesGroupingSeparator = true
        formatter.minimumFractionDigits = 4
        formatter.maximumFractionDigits = 4
        formatter.minimumIntegerDigits = 1
        return formatter.string(from: NSNumber(value: value)) ?? String(format: "%.4f", value)
    }

    private static func decimalizeIntegerTokens(in expression: String) -> String {
        let pattern = "(?<![A-Za-z0-9_\\.])([0-9]+)(?![A-Za-z0-9_\\.])"
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return expression }

        let range = NSRange(expression.startIndex..<expression.endIndex, in: expression)
        let matches = regex.matches(in: expression, range: range)
        var output = expression
        for match in matches.reversed() {
            guard let tokenRange = Range(match.range(at: 1), in: output) else { continue }
            output.replaceSubrange(tokenRange, with: output[tokenRange] + ".0")
        }
        return output
    }

    private static func normalizeExpression(_ expression: String) -> String {
        let normalized = expression
            .replacingOccurrences(of: "x", with: "*")
            .replacingOccurrences(of: "X", with: "*")
            .replacingOccurrences(of: ":", with: "/")
        return replacePrefixSqrt(in: normalized)
    }

    private static func replacePrefixSqrt(in expression: String) -> String {
        var output = ""
        var index = expression.startIndex

        while index < expression.endIndex {
            let char = expression[index]
            if char == "v" || char == "V" {
                let prev = index > expression.startIndex ? expression[expression.index(before: index)] : " "
                let nextIndex = expression.index(after: index)
                let next = nextIndex < expression.endIndex ? expression[nextIndex] : " "
                let prevIsWord = prev.isLetter || prev.isNumber || prev == "_"
                let nextIsStart = next.isNumber || next == "." || next == "(" || next == " "
                if !prevIsWord && nextIsStart {
                    output.append("sqrt")
                    index = nextIndex
                    continue
                }
            }
            output.append(char)
            index = expression.index(after: index)
        }
        return output
    }
}

private enum ParserError: Error {
    case invalidExpression
    case divisionByZero
}

private struct Parser {
    private let chars: [Character]
    private var index: Int = 0

    init(input: String) {
        self.chars = Array(input)
    }

    mutating func parse() throws -> Double {
        let value = try parseExpression()
        skipWhitespace()
        guard index == chars.count else {
            throw ParserError.invalidExpression
        }
        return value
    }

    private mutating func parseExpression() throws -> Double {
        var value = try parseTerm()
        while true {
            skipWhitespace()
            if consume("+") {
                value += try parseTerm()
            } else if consume("-") {
                value -= try parseTerm()
            } else {
                return value
            }
        }
    }

    private mutating func parseTerm() throws -> Double {
        var value = try parseUnary()
        while true {
            skipWhitespace()
            if consume("*") {
                value *= try parseUnary()
            } else if consume("/") {
                let divisor = try parseUnary()
                if divisor == 0 {
                    throw ParserError.divisionByZero
                }
                value /= divisor
            } else if consume("%") {
                let divisor = try parseUnary()
                if divisor == 0 {
                    throw ParserError.divisionByZero
                }
                value = value.truncatingRemainder(dividingBy: divisor)
            } else {
                return value
            }
        }
    }

    private mutating func parsePower() throws -> Double {
        var value = try parsePrimary()
        skipWhitespace()
        if consume("^") {
            let exponent = try parseUnary()
            value = Foundation.pow(value, exponent)
            if value.isNaN || value.isInfinite {
                throw ParserError.invalidExpression
            }
        }
        return value
    }

    private mutating func parseUnary() throws -> Double {
        skipWhitespace()

        if consume("+") {
            return try parseUnary()
        }
        if consume("-") {
            return -(try parseUnary())
        }

        if matchKeyword("sqrt") {
            _ = consumeKeyword("sqrt")
            return try applyFunction("sqrt", to: try parseFunctionArgument())
        }

        if matchKeyword("abs") {
            _ = consumeKeyword("abs")
            return try applyFunction("abs", to: try parseFunctionArgument())
        }

        if matchKeyword("round") {
            _ = consumeKeyword("round")
            return try applyFunction("round", to: try parseFunctionArgument())
        }

        if matchKeyword("floor") {
            _ = consumeKeyword("floor")
            return try applyFunction("floor", to: try parseFunctionArgument())
        }

        if matchKeyword("ceil") {
            _ = consumeKeyword("ceil")
            return try applyFunction("ceil", to: try parseFunctionArgument())
        }

        return try parsePower()
    }

    private mutating func parsePrimary() throws -> Double {
        skipWhitespace()

        if consume("(") {
            let value = try parseExpression()
            skipWhitespace()
            guard consume(")") else {
                throw ParserError.invalidExpression
            }
            return try applyPostfixOperators(to: value)
        }

        if index < chars.count, chars[index].isLetter {
            let ident = parseIdentifier()
            switch ident.lowercased() {
            case "pi":
                return try applyPostfixOperators(to: .pi)
            case "e":
                return try applyPostfixOperators(to: M_E)
            default:
                throw ParserError.invalidExpression
            }
        }

        let number = try parseNumber()
        return try applyPostfixOperators(to: number)
    }

    private mutating func applyPostfixOperators(to seed: Double) throws -> Double {
        var value = seed
        while true {
            skipWhitespace()
            if consume("!") {
                value = try factorial(value)
                continue
            }
            if shouldConsumePostfixPercent() {
                _ = consume("%")
                value /= 100
                continue
            }
            return value
        }
    }

    private mutating func shouldConsumePostfixPercent() -> Bool {
        guard index < chars.count, chars[index] == "%" else { return false }
        var lookahead = index + 1
        while lookahead < chars.count && chars[lookahead].isWhitespace {
            lookahead += 1
        }
        guard lookahead < chars.count else { return true }
        let next = chars[lookahead]
        if next.isNumber || next == "." || next == "(" || next.isLetter {
            return false
        }
        return true
    }

    private mutating func parseFunctionArgument() throws -> Double {
        skipWhitespace()
        if consume("(") {
            let value = try parseExpression()
            skipWhitespace()
            guard consume(")") else { throw ParserError.invalidExpression }
            return value
        }
        return try parseUnary()
    }

    private func applyFunction(_ name: String, to value: Double) throws -> Double {
        switch name {
        case "sqrt":
            guard value >= 0 else { throw ParserError.invalidExpression }
            return Foundation.sqrt(value)
        case "abs":
            return Swift.abs(value)
        case "round":
            return Foundation.round(value)
        case "floor":
            return Foundation.floor(value)
        case "ceil":
            return Foundation.ceil(value)
        default:
            throw ParserError.invalidExpression
        }
    }

    private func factorial(_ value: Double) throws -> Double {
        guard value >= 0, value.rounded() == value else {
            throw ParserError.invalidExpression
        }
        let n = Int(value)
        guard n <= 170 else {
            throw ParserError.invalidExpression
        }
        if n <= 1 { return 1 }
        return (2...n).reduce(1.0) { $0 * Double($1) }
    }

    private mutating func parseIdentifier() -> String {
        let start = index
        while index < chars.count, chars[index].isLetter || chars[index] == "_" {
            index += 1
        }
        return String(chars[start..<index])
    }

    private mutating func parseNumber() throws -> Double {
        skipWhitespace()
        let start = index
        var sawDigit = false
        var sawDot = false

        while index < chars.count {
            let ch = chars[index]
            if ch.isNumber {
                sawDigit = true
                index += 1
            } else if ch == "." && !sawDot {
                sawDot = true
                index += 1
            } else {
                break
            }
        }

        guard sawDigit else {
            throw ParserError.invalidExpression
        }

        let token = String(chars[start..<index])
        guard let value = Double(token) else {
            throw ParserError.invalidExpression
        }
        return value
    }

    private mutating func skipWhitespace() {
        while index < chars.count && chars[index].isWhitespace {
            index += 1
        }
    }

    private mutating func consume(_ ch: Character) -> Bool {
        guard index < chars.count, chars[index] == ch else { return false }
        index += 1
        return true
    }

    private func matchKeyword(_ keyword: String) -> Bool {
        let end = index + keyword.count
        guard end <= chars.count else { return false }
        let token = String(chars[index..<end])
        return token.lowercased() == keyword
    }

    private mutating func consumeKeyword(_ keyword: String) -> Bool {
        guard matchKeyword(keyword) else { return false }
        index += keyword.count
        return true
    }
}

File: apps/windows/LauncherApp/Commands/CalcCommand.cs
using System;
using System.Globalization;

namespace LauncherApp.Commands;

public static class CalcCommand
{
    private const double MaxMagnitude = 1_000_000_000_000.0;

    public static bool TryEvaluate(string expression, out string message)
    {
        message = "Invalid expression";
        if (!IsReadyForEvaluation(expression))
        {
            return false;
        }

        string normalized = NormalizeExpression(expression);
        try
        {
            var parser = new Parser(normalized);
            double value = parser.Parse();

            if (Math.Abs(value) > MaxMagnitude)
            {
                message = "Error: result out of range (+/-1,000,000,000,000)";
                return false;
            }

            message = $"Result: {FormatFloat(value)}";
            return true;
        }
        catch (DivideByZeroException)
        {
            message = "Error: division by zero";
            return false;
        }
        catch
        {
            return false;
        }
    }

    public static bool IsReadyForEvaluation(string expression)
    {
        if (string.IsNullOrWhiteSpace(expression))
        {
            return false;
        }

        string trimmed = expression.Trim();
        int balance = 0;
        foreach (char ch in trimmed)
        {
            if (!IsAllowedChar(ch))
            {
                return false;
            }

            if (ch == '(')
            {
                balance++;
            }
            else if (ch == ')')
            {
                balance--;
                if (balance < 0)
                {
                    return false;
                }
            }
        }

        if (balance != 0)
        {
            return false;
        }

        char last = trimmed[^1];
        if ("+-*/%^.(".IndexOf(last) >= 0)
        {
            return false;
        }

        return true;
    }

    private static bool IsAllowedChar(char ch)
    {
        return char.IsLetterOrDigit(ch)
            || ch == '_'
            || ch == '+'
            || ch == '-'
            || ch == '*'
            || ch == '/'
            || ch == '%'
            || ch == '^'
            || ch == '!'
            || ch == '('
            || ch == ')'
            || ch == '.'
            || ch == ':'
            || ch == 'x'
            || ch == 'X'
            || ch == 'v'
            || ch == 'V'
            || char.IsWhiteSpace(ch);
    }

    private static string NormalizeExpression(string expression)
    {
        string normalized = expression
            .Trim()
            .Replace('x', '*')
            .Replace('X', '*')
            .Replace(':', '/');

        return ReplacePrefixSqrt(normalized);
    }

    private static string ReplacePrefixSqrt(string expression)
    {
        var output = new System.Text.StringBuilder(expression.Length + 4);
        for (int i = 0; i < expression.Length; i++)
        {
            char current = expression[i];
            if (current == 'v' || current == 'V')
            {
                char prev = i > 0 ? expression[i - 1] : ' ';
                char next = i + 1 < expression.Length ? expression[i + 1] : ' ';
                bool prevIsWord = char.IsLetterOrDigit(prev) || prev == '_';
                bool nextIsStart = char.IsDigit(next) || next == '.' || next == '(' || char.IsWhiteSpace(next);

                if (!prevIsWord && nextIsStart)
                {
                    output.Append("sqrt");
                    continue;
                }
            }

            output.Append(current);
        }

        return output.ToString();
    }

    private static string FormatFloat(double value)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            return "nan";
        }

        var format = (NumberFormatInfo)CultureInfo.GetCultureInfo("en-US").NumberFormat.Clone();
        format.NumberGroupSeparator = ",";
        return value.ToString("N4", format);
    }

    private sealed class Parser
    {
        private readonly struct ValueNode
        {
            public double Value { get; }
            public bool IsStandalonePercent { get; }
            public double PercentFraction { get; }

            public ValueNode(double value, bool isStandalonePercent = false, double percentFraction = 0)
            {
                Value = value;
                IsStandalonePercent = isStandalonePercent;
                PercentFraction = percentFraction;
            }
        }

        private readonly char[] _chars;
        private int _index;

        public Parser(string input)
        {
            _chars = input.ToCharArray();
        }

        public double Parse()
        {
            ValueNode value = ParseExpression();
            SkipWhitespace();
            if (_index != _chars.Length)
            {
                throw new InvalidOperationException("Invalid expression");
            }

            return value.Value;
        }

        private ValueNode ParseExpression()
        {
            ValueNode value = ParseTerm();
            while (true)
            {
                SkipWhitespace();
                if (Consume('+'))
                {
                    ValueNode rhs = ParseTerm();
                    double combined = rhs.IsStandalonePercent
                        ? value.Value + (value.Value * rhs.PercentFraction)
                        : value.Value + rhs.Value;
                    value = new ValueNode(combined);
                }
                else if (Consume('-'))
                {
                    ValueNode rhs = ParseTerm();
                    double combined = rhs.IsStandalonePercent
                        ? value.Value - (value.Value * rhs.PercentFraction)
                        : value.Value - rhs.Value;
                    value = new ValueNode(combined);
                }
                else
                {
                    return value;
                }
            }
        }

        private ValueNode ParseTerm()
        {
            ValueNode value = ParseUnary();
            while (true)
            {
                SkipWhitespace();
                if (Consume('*'))
                {
                    ValueNode rhs = ParseUnary();
                    value = new ValueNode(value.Value * rhs.Value);
                }
                else if (Consume('/'))
                {
                    ValueNode rhs = ParseUnary();
                    double divisor = rhs.Value;
                    if (divisor == 0)
                    {
                        throw new DivideByZeroException();
                    }

                    value = new ValueNode(value.Value / divisor);
                }
                else if (Consume('%'))
                {
                    ValueNode rhs = ParseUnary();
                    double divisor = rhs.Value;
                    if (divisor == 0)
                    {
                        throw new DivideByZeroException();
                    }

                    value = new ValueNode(value.Value % divisor);
                }
                else
                {
                    return value;
                }
            }
        }

        private ValueNode ParsePower()
        {
            ValueNode value = ParsePrimary();
            SkipWhitespace();
            if (Consume('^'))
            {
                ValueNode exponent = ParseUnary();
                double powered = Math.Pow(value.Value, exponent.Value);
                if (double.IsNaN(powered) || double.IsInfinity(powered))
                {
                    throw new InvalidOperationException("Invalid expression");
                }

                value = new ValueNode(powered);
            }

            return value;
        }

        private ValueNode ParseUnary()
        {
            SkipWhitespace();

            if (Consume('+'))
            {
                return ParseUnary();
            }

            if (Consume('-'))
            {
                ValueNode negated = ParseUnary();
                return new ValueNode(-negated.Value);
            }

            if (MatchKeyword("sqrt"))
            {
                ConsumeKeyword("sqrt");
                return new ValueNode(ApplyFunction("sqrt", ParseFunctionArgument()));
            }

            if (MatchKeyword("abs"))
            {
                ConsumeKeyword("abs");
                return new ValueNode(ApplyFunction("abs", ParseFunctionArgument()));
            }

            if (MatchKeyword("round"))
            {
                ConsumeKeyword("round");
                return new ValueNode(ApplyFunction("round", ParseFunctionArgument()));
            }

            if (MatchKeyword("floor"))
            {
                ConsumeKeyword("floor");
                return new ValueNode(ApplyFunction("floor", ParseFunctionArgument()));
            }

            if (MatchKeyword("ceil"))
            {
                ConsumeKeyword("ceil");
                return new ValueNode(ApplyFunction("ceil", ParseFunctionArgument()));
            }

            return ParsePower();
        }

        private ValueNode ParsePrimary()
        {
            SkipWhitespace();

            if (Consume('('))
            {
                ValueNode value = ParseExpression();
                SkipWhitespace();
                if (!Consume(')'))
                {
                    throw new InvalidOperationException("Invalid expression");
                }

                return ApplyPostfixOperators(value.Value);
            }

            if (_index < _chars.Length && char.IsLetter(_chars[_index]))
            {
                string ident = ParseIdentifier();
                double constant = ident.ToLowerInvariant() switch
                {
                    "pi" => Math.PI,
                    "e" => Math.E,
                    _ => throw new InvalidOperationException("Invalid expression"),
                };

                return ApplyPostfixOperators(constant);
            }

            double number = ParseNumber();
            return ApplyPostfixOperators(number);
        }

        private ValueNode ApplyPostfixOperators(double seed)
        {
            var node = new ValueNode(seed);
            while (true)
            {
                SkipWhitespace();
                if (Consume('!'))
                {
                    node = new ValueNode(Factorial(node.Value));
                    continue;
                }

                if (ShouldConsumePostfixPercent())
                {
                    Consume('%');
                    double fraction = node.Value / 100d;
                    node = new ValueNode(fraction, isStandalonePercent: true, percentFraction: fraction);
                    continue;
                }

                return node;
            }
        }

        private bool ShouldConsumePostfixPercent()
        {
            if (_index >= _chars.Length || _chars[_index] != '%')
            {
                return false;
            }

            int lookahead = _index + 1;
            while (lookahead < _chars.Length && char.IsWhiteSpace(_chars[lookahead]))
            {
                lookahead++;
            }

            if (lookahead >= _chars.Length)
            {
                return true;
            }

            char next = _chars[lookahead];
            if (char.IsDigit(next) || next == '.' || next == '(' || char.IsLetter(next))
            {
                return false;
            }

            return true;
        }

        private double ParseFunctionArgument()
        {
            SkipWhitespace();
            if (Consume('('))
            {
                ValueNode value = ParseExpression();
                SkipWhitespace();
                if (!Consume(')'))
                {
                    throw new InvalidOperationException("Invalid expression");
                }

                return value.Value;
            }

            return ParseUnary().Value;
        }

        private static double ApplyFunction(string name, double value)
        {
            return name switch
            {
                "sqrt" when value < 0 => throw new InvalidOperationException("Invalid expression"),
                "sqrt" => Math.Sqrt(value),
                "abs" => Math.Abs(value),
                "round" => Math.Round(value),
                "floor" => Math.Floor(value),
                "ceil" => Math.Ceiling(value),
                _ => throw new InvalidOperationException("Invalid expression"),
            };
        }

        private static double Factorial(double value)
        {
            if (value < 0 || Math.Round(value) != value)
            {
                throw new InvalidOperationException("Invalid expression");
            }

            int n = (int)value;
            if (n > 170)
            {
                throw new InvalidOperationException("Invalid expression");
            }

            if (n <= 1)
            {
                return 1;
            }

            double result = 1;
            for (int i = 2; i <= n; i++)
            {
                result *= i;
            }

            return result;
        }

        private string ParseIdentifier()
        {
            int start = _index;
            while (_index < _chars.Length && (char.IsLetter(_chars[_index]) || _chars[_index] == '_'))
            {
                _index++;
            }

            return new string(_chars[start.._index]);
        }

        private double ParseNumber()
        {
            SkipWhitespace();
            int start = _index;
            bool sawDigit = false;
            bool sawDot = false;

            while (_index < _chars.Length)
            {
                char ch = _chars[_index];
                if (char.IsDigit(ch))
                {
                    sawDigit = true;
                    _index++;
                }
                else if (ch == '.' && !sawDot)
                {
                    sawDot = true;
                    _index++;
                }
                else
                {
                    break;
                }
            }

            if (!sawDigit)
            {
                throw new InvalidOperationException("Invalid expression");
            }

            string token = new(_chars[start.._index]);
            if (!double.TryParse(token, NumberStyles.Float, CultureInfo.InvariantCulture, out double value))
            {
                throw new InvalidOperationException("Invalid expression");
            }

            return value;
        }

        private void SkipWhitespace()
        {
            while (_index < _chars.Length && char.IsWhiteSpace(_chars[_index]))
            {
                _index++;
            }
        }

        private bool Consume(char ch)
        {
            if (_index >= _chars.Length || _chars[_index] != ch)
            {
                return false;
            }

            _index++;
            return true;
        }

        private bool MatchKeyword(string keyword)
        {
            int end = _index + keyword.Length;
            if (end > _chars.Length)
            {
                return false;
            }

            for (int i = 0; i < keyword.Length; i++)
            {
                if (char.ToLowerInvariant(_chars[_index + i]) != keyword[i])
                {
                    return false;
                }
            }

            return true;
        }

        private bool ConsumeKeyword(string keyword)
        {
            if (!MatchKeyword(keyword))
            {
                return false;
            }

            _index += keyword.Length;
            return true;
        }
    }
}
